specs/accounting_reform/beacon-chain.md

```diff
@@ -0,0 +1,285 @@
+    balances: List[Gwei, VALIDATOR_REGISTRY_LIMIT]
+    # Randomness
+    randao_mixes: Vector[Bytes32, EPOCHS_PER_HISTORICAL_VECTOR]
+    # Slashings
+    slashings: Vector[Gwei, EPOCHS_PER_SLASHINGS_VECTOR]  # Per-epoch sums of slashed effective balances
+    # Current and previous epoch participation flags
+    previous_epoch_flags: List[Bitvector[8], VALIDATOR_REGISTRY_LIMIT]
+    current_epoch_flags: List[Bitvector[8], VALIDATOR_REGISTRY_LIMIT]
+    # Finality
+    justification_bits: Bitvector[JUSTIFICATION_BITS_LENGTH]  # Bit set for every recent justified epoch
+    previous_justified_checkpoint: Checkpoint
+    current_justified_checkpoint: Checkpoint
+    finalized_checkpoint: Checkpoint
+```
+
+###  Helpers
+
+##### `get_base_reward`
+
+```python
+def get_base_reward(state: BeaconState, index: ValidatorIndex) -> Gwei:
+    total_balance = get_total_active_balance(state)
+    effective_balance = state.validators[index].effective_balance
+    return Gwei(effective_balance * BASE_REWARD_FACTOR // integer_squareroot(total_balance))
+```
+
+Note: remove the `// BASE_REWARDS_PER_EPOCH`
+
+##### `process_attestation`
+
+```python
+def process_attestation(state: BeaconState, attestation: Attestation) -> None:
+    data = attestation.data
+    assert data.target.epoch in (get_previous_epoch(state), get_current_epoch(state))
+    assert data.target.epoch == compute_epoch_at_slot(data.slot)
+    assert data.slot + MIN_ATTESTATION_INCLUSION_DELAY <= state.slot <= data.slot + SLOTS_PER_EPOCH
+    assert data.index < get_committee_count_per_slot(state, data.target.epoch)
+
+    committee = get_beacon_committee(state, data.slot, data.index)
+    assert len(attestation.aggregation_bits) == len(committee)
+        
+    if data.target.epoch == get_current_epoch(state):
+        flags = state.current_epoch_flags
+        assert data.source == state.current_justified_checkpoint
+    else:
+        flags = state.previous_epoch_flags
+        assert data.source == state.previous_justified_checkpoint    
+
+    # Verify signature
+    assert is_valid_indexed_attestation(state, get_indexed_attestation(state, attestation))
+        
+    # Process flags
+    flags_to_set = []
+    if data.target.root == get_block_root(state, data.target.epoch):
+        flags_to_set.append(FLAG_TARGET)
+
+        is_correct_head = data.beacon_block_root == get_block_root_at_slot(state, data.slot)
+        is_very_timely = state.slot == data.slot + MIN_ATTESTATION_INCLUSION_DELAY
+        if is_correct_head and is_very_timely:
+            flags_to_set.append(FLAG_HEAD_AND_VERY_TIMELY)
+
+    if state.slot <= data.slot + integer_squareroot(SLOTS_PER_EPOCH):
+        flags_to_set.append(FLAG_TIMELY)
+
+    # Update participation flags
+    for participant in get_attesting_indices(state, data, attestation.aggregation_bits):
+        active_position = get_active_validator_indices(state, data.target.epoch).index(participant)
```

Comment:
Take pyspec for example, it seems that `state` object will be changed when setting `flags[active_position][flag] = True` with mutable side effect and the LRU cache keys of `get_active_validator_indices` will be different for each `participant`.

A slight refactoring may make implementers happier:

```suggestion
    active_validator_indices = get_active_validator_indices(state, data.target.epoch)
    for participant in get_attesting_indices(state, data, attestation.aggregation_bits):
        active_position = active_validator_indices.index(participant)
```